doc/src/guides/contributing/deprecations.md

```diff
@@ -0,0 +1,388 @@
+(deprecation-policy)=
+# Deprecation Policy
+
+This page outlines SymPy's police on doing deprecations, and describes the
+steps developers should take to properly deprecate code.
+
+## What is a deprecation?
+
+A deprecation is a way to make backwards incompatible changes in a way that
+allows users to update their code. Deprecated code continues to work as it
+used to, but whenever someone uses it, it prints a warning to the screen
+indicating that it will be removed in a future version of SymPy, and
+indicating what the user should be using instead.
+
+This allows users a chance to update their code without it completely
+breaking. It also gives SymPy an opportunity to give users an informative
+message on how to update their code, rather than making their code simply
+error or start giving wrong answers.
+
+## When to deprecate and when not to
+
+Sometimes the API of SymPy must change in an incompatible way. We try to avoid
+this, because changing API breaks people's code, but it is often deemed worthy
+to do so. Some reasons APIs are changed can include:
+
+- The existing API is confusing.
+- There is unnecessary redundancy in the API.
+- The existing API limits what is possible.
+
+Because one of the core use-cases of SymPy is to be usable as a library, we
+take API breakage very seriously. Whenever an API breakage is necessary, the
+following steps should be taken:
+
+- Discuss the API change with the community. Be sure that the improved API is
+  indeed better, and worth a breakage. It is important to get API right so
+  that things will not need to break again to "fix" it a second time.
+- Document any API breakages in the relevant section of the release notes for
+  the relevant release. This section is typically near the top of the release
+  notes.
+- If it is possible, deprecate the old API.
+
+When considering whether a change requires a deprecation, two things must be
+considered:
+
+- Is the change backwards incompatible?
+- Is the behavior being changed public API?
+
+A change is backwards incompatible if user code making use of it would stop
+working after the change.
+
+What counts as "public API" needs to be considered on a case-by-case basis.
+The exact rules for what does and doesn't constitute public API for SymPy
+are still not fully codified. Very broadly, some thing that constitute public
+API are (this list is not exhaustive)
+
+- Function names.
+- Keyword argument names.
+- Keyword argument default values.
+- The mathematical conventions used to define a function.
+- Submodule names for **public** functions that are not included in the
+  top-level `sympy/__init__.py`, since users must use the fully qualified
+  module name to import them (note: many such functions are not included
+  because they aren't actually public API at all. For these, no deprecations
+  are required to make breaking changes).
```

Comment:
I don't really agree with this one. `sympy.physics.vector.frame.ReferenceFrame` isn't importable like `from sympy import ReferenceFrame`, but it is certainly public API. Does this bullet say that only names imported by `from sympy import *` are public?

Reply:
There needs to be a more nuanced definition here. For most of the codebase it is true that public API is exposed at top-level like `from sympy import X` but there are major exceptions like physics, stats, combinatorics...

Reply:
I think you aren't understanding it. What I am trying to write here agrees with you. 

The basic point is this: if users *have* to write the module name to access a function (because it can't be imported from the top-level `from sympy import X`), then the module name itself is part of the public API. This is because if we renamed the module, it would break user code. 

OTOH if a function is in the top-level, then users can write `from sympy import X` instead of `from sympy.submodule.submodule import X`, so if we rename `submodule` it won't break user code. 

Or at least that's the ideal situation. It doesn't consider users who may be importing top-level names directly from submodules anyway. I've actually seen quite a bit of user code that does things like `from sympy.functions import exp` instead of `from sympy import exp`. The fact that we intentionally do this in library code doesn't help here. So in practice, if we ever did want to rename modules or move functions around, we might want to do a deprecation anyway. Maintaining backwards compatibility here is quite trivial (see for instance the deprecation for `sympy.printing.ccode`, which was done for technical reasons). 

I think a more codified public/private API policy should probably include things like underscores for names that aren't supposed to be public. So e.g., `sympy.functions` would become `sympy._functions` and all user code should import names from there from the top-level. But this requires a *lot* of work to actually do, so it's quite out of scope for this PR. For now, I just want to give some general guidelines. 

I'll reword these two bullet points to make them clearer.

Reply:
I guess I'll just remove the non-public half of this. The problem is that sometimes names are in a file that aren't included in `__init__.py` but are still public API. The most common example of this is base classes, which we want people to be able to subclass. To really distinguish this requires a lot more work. 

The cleanest thing I can say for now is that submodule renames should be deprecated (which as I noted, is not difficult). Even if a function is just moved around, it's very easy to keep an import of it at the top of the file so that direct imports from the old module still work.

Reply:
See what you think of it now.

Also it really needs to be made clear here that these lists here are just guidelines. We don't have a real public/private API policy yet. There are always exceptions to the rules.